### PR TITLE
Publish richer virtual disk statistics to oximeter

### PR DESCRIPTION
**Pull request by bnaecker**
- Use the `virtual_disk:*` timeseries defined in Omicron to publish statistics about the block operations performed on virtual disks.
- Add concept of a completion callback to the existing block tracking object, and expose a way to set the callback on a `block::Device`.
- Create object for managing all the statistics for one virtual disk backed by Crucible. This is shared between the existing oximeter producer, and the block devices, via the above completion callback mechanism. As I/Os complete, the stats are updated, and oximeter collects them periodically.
- Straighten out a bunch of the initialiazation logic for the pre-existing metrics, producer server, and registry. This grew pretty organically, and was getting unwieldy and confusing.
- Use `virtual_machine:reset` definition from TOML
- Move kstat sampler creation to machine initialization process, rather than buried in the producer server startup. This lets us use the sampler in-line when initializing the vCPUs, where it should be much easier to re-use when we want to add more kstat-based metrics, such as Viona link stats.
- Simplify and cleanup the `ServerStats` types, and remove the kstat sampler from it, which never should have been there at all.
- Bump kstat-rs dep, which includes the new version that is cross-platform. This gets rid of a bunch of cfg-soup in the stats modules.

### CODE REVIEW

**Comment by bnaecker** [comment]
This still needs some testing, but it's a larger chunk than I intended so I'd like to get early eyes on it.  I took this opportunity to rework a lot of the metrics initialization, which had become pretty Byzantine.

The main point here is to publish richer virtual disk metrics than what we currently get from Crucible. This adds the same counters we had before, and also failure counters and histograms tracking I/O latency and I/O sizes. These use the same block device tracking machinery that was added to handle USDT probes, adding a completion callback to that struct. We now set the completion callback to update the statistics for each virtual disk. Those stats are shared with the oximeter producer registry, so they're published along with everything else.

The extras that came along for the ride here are a mishmash of cleanup tasks, mostly listed in the commit message. This PR also relies on a forthcoming PR in Omicron that adds the schema for the new statistics. I'll link that once I post it, and follow up with testing notes as well.

File: bin/propolis-server/src/lib/stats/virtual_disk.rs
Comment (leftwo) [inline bin/propolis-server/src/lib/stats/virtual_disk.rs:213]: The `sled_id` here is going to be the sled the disk was attached to when it was in use, right?
We don't want to confuse that with the location(s) of all the downstairs that could be what underpins the virtual disk that Propolis sees.

File: lib/propolis/src/block/crucible.rs
Comment (leftwo) [inline lib/propolis/src/block/crucible.rs:159]: I'm not sure what happens if we call this before the upstairs is active.
When you run this, do you get a block size back from the volume?
If it works, then you are calling it after activation or I don't know what I'm talking about.  Possibly both.

File: bin/propolis-server/src/lib/stats/virtual_disk.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/stats/virtual_disk.rs:213]: That's right, it's the ID of the sled hosting the instance. I think the description in the `virtual-disk.toml` schema should make this clear, but let me know if that's lacking in some way.

File: lib/propolis/src/block/crucible.rs
Comment (bnaecker) [reply lib/propolis/src/block/crucible.rs:159]: Welp, I'm about to test it, so I'll let you know! We also get the block size in the API request from the sled-agent, so it's possible we can pass it down from there to where we need it in another way, rather than asking Crucible.

**Comment by bnaecker** [comment]
Talked offline with @ahl. He questions the value of putting the sled identifiers on this kind of timeseries at all, which I can understand. I might have been overzealous in my attempt to address https://github.com/oxidecomputer/omicron/issues/5267.

**Comment by bnaecker** [comment]
Ok, I've done an initial round of testing. The Omicron side is still blocked on https://github.com/oxidecomputer/omicron/pull/6409 and this WIP branch: https://github.com/oxidecomputer/omicron/tree/new-virtual-disk-timeseries-definition. But incorporating both of those locally and creating a VM does result in the expected timeseries in ClickHouse.

Here are all of them:

```
0x〉\l
oximeter_collector:collections
virtual_disk:bytes_read
virtual_disk:bytes_written
virtual_disk:failed_flushes
virtual_disk:failed_reads
virtual_disk:failed_writes
virtual_disk:flushes
virtual_disk:io_latency
virtual_disk:io_size
virtual_disk:reads
virtual_disk:writes
virtual_machine:pv_panic_guest_handled
virtual_machine:pv_panic_host_handled
virtual_machine:reset
virtual_machine:vcpu_usage
```

Digging into individual things looks like I expect:

```
0x〉get virtual_disk:reads | last 1

virtual_disk:reads

 attached_instance_id: 6d39bc42-74d3-4f34-a035-6a068a4e4da3
 block_size: 4096
 disk_id: d82074ff-7098-483f-90f3-d7fca35baba6
 project_id: d50c3c4d-0008-45a3-a285-3b9c8ef84d38
 silo_id: 7fec5b38-3c43-4ed1-ab05-5152401c46a8
 sled_id: faf169b9-bf53-4a53-8f52-68c94dc377bd
 sled_model: fake-gimlet
 sled_revision: 1
 sled_serial: fake-serial
   [2024-08-22 21:44:10.723454479, 2024-08-22 21:54:40.861305986]: [109]
0x〉get virtual_disk:io_size | filter io_kind == "read" | last 1

virtual_disk:io_size

 attached_instance_id: 6d39bc42-74d3-4f34-a035-6a068a4e4da3
 block_size: 4096
 disk_id: d82074ff-7098-483f-90f3-d7fca35baba6
 io_kind: read
 project_id: d50c3c4d-0008-45a3-a285-3b9c8ef84d38
 silo_id: 7fec5b38-3c43-4ed1-ab05-5152401c46a8
 sled_id: faf169b9-bf53-4a53-8f52-68c94dc377bd
 sled_model: fake-gimlet
 sled_revision: 1
 sled_serial: fake-serial
   [2024-08-22 21:44:10.723879738, 2024-08-22 21:55:10.862958317]: [0: 105, 4096: 1, 8192: 3, 16384: 0, 32768: 0, 65536: 0, 131072: 0, 262144: 0, 524288: 0, 1048576: 0, 2097152: 0, 4194304: 0, 8388608: 0, 16777216: 0, 33554432: 0, 67108864: 0, 134217728: 0, 268435456: 0, 536870912: 0, 1073741824: 0, min: 512, max: 8192, mean: 1211.8899082568807, std_dev: 1407.1530457991503, p50: 748.4546500654893, p90: 2197.500566706399, p99: 6740.062811662923]
```

So that shows we had 109 total reads, and the I/O size histogram shows that those were broken out into 105 in the range [0, 4KiB), 1 in [4KiB, 8KiB), and 3 in [8KiB, 16KiB). So there are the same number of items in both, meaning we're consistent at least.

File: Cargo.toml
Comment (bnaecker) [inline Cargo.toml:74]: Note to self: Point back to Omicron main when https://github.com/oxidecomputer/omicron/pull/6420 merges

File: Cargo.toml
Comment (bnaecker) [inline Cargo.toml:84]: Note to self: Point back to rev on Crucible main when https://github.com/oxidecomputer/crucible/pull/1429 merges

**Comment by bnaecker** [comment]
There's one kind of structural question my testing revealed. These virtual disk stats are updated all the time, but only sampled when `oximeter` asks for it. That means we get samples every 30s or 10s or whatever we pick. We could decouple these, by sampling the stats locally more frequently, and publishing more than one sample's worth of data to `oximeter`. That does require a different architecture though, since we'd need a small polling loop pushing data to a queue somewhere.

I think it largely depends on how densely we want to sample the data. I'd defer to @leftwo and @ahl on that probably. I reduced the collection interval to 10s, which still feels borderline, but I admit that's a WAG.

File: bin/propolis-server/src/lib/vm/ensure.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/vm/ensure.rs:187]: Is it worthwhile to create the sampler if the `producer_registry` in the `options` is `None`? (In particular, does the `kstat_sampler`'s memory usage grow over time if it never gets collected from? We'll have a producer in production Propolises, of course, but it's sometimes useful to be able to run long-running tests with a freestanding propolis-server binary.)

File: bin/propolis-server/src/lib/initializer.rs
Comment (leftwo) [inline bin/propolis-server/src/lib/initializer.rs:613]: If we can't register stats, will this bail out of the loop early?

It seems like we are in the loop:
`for (name, device_spec) in &self.spec.devices.storage_devices {`

And, if we `retrun Ok`, will the caller even know that not all of their devices were not handled.

The same is true about if we can't get a block size above, we could bail early and not attach other disks.

File: bin/propolis-server/src/lib/stats.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/stats.rs:59]: I was expecting it to go away as well. But the `Reset` metric _is_ currently published to `oximeter`, so I think some form of this needs to stick around. At that point, it's similar to the current block stats: there's some state shared between the Propolis server, which updates the reset count, and the `oximeter_producer::Server`, which publishes the value when `oximeter` polls for it.

It's possible it could be shared differently, so I'll take another look at it.

**Review by leftwo — COMMENTED** [review]
Does this leave in place the statistics that the crucible upstairs logs as well?

File: bin/propolis-server/src/lib/stats.rs
Comment (gjcolombo) [reply bin/propolis-server/src/lib/stats.rs:59]: Thanks! Don't spend too much time on it; I was mostly just curious about how this had shaken out in practice.

File: bin/propolis-server/src/lib/vm/ensure.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/vm/ensure.rs:187]: The `create_kstat_sampler()` method returns an `Option<KstatSampler>`. That's `None` if either the `producer_registry` is `None`, or if we just fail to create the thing. I'll rename this variable to make that more clear!

That obviates your second question, about memory consumption, but the answer is "no". The `KstatSampler` requires one to choose a limited buffer size, used in a ringbufffer to maintain uncollected samples. The memory consumption is bounded, even if it's never polled for data.

File: bin/propolis-server/src/lib/initializer.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/initializer.rs:613]: Nice catch! This is definitely wrong, we should be `continue`ing here instead. I'll update that, thanks.

File: bin/propolis-server/src/lib/vm/ensure.rs
Comment (gjcolombo) [reply bin/propolis-server/src/lib/vm/ensure.rs:187]: That makes sense--thanks for the explanation!

**Comment by bnaecker** [comment]
> Does this leave in place the statistics that the crucible upstairs logs as well?

Yeah, the existing `crucible_upstairs:*` stats are still there. Those are currently used for the web console's view of instance stats. I think we should switch that over to these data, and then we can separately decide to remove the original data if we want.

File: bin/propolis-server/src/lib/initializer.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/initializer.rs]: Done in 15327f2

File: bin/propolis-server/src/lib/initializer.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/initializer.rs:615]: Switched to a `continue` in 15327f2

File: bin/propolis-server/src/lib/initializer.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/initializer.rs]: Done in 15327f2cd7d9401c4de97c8dda4c4afff37f04d1

File: bin/propolis-server/src/lib/stats.rs
Comment (bnaecker) [reply bin/propolis-server/src/lib/stats.rs:59]: I think this is in a reasonable place as-is. Cleaning up the naming and the initialization process makes it much clearer that the `ServerStats` type is used to store metrics about the API server or requests it services. We could add HTTP latency tracking here for example, if we cared to know that. Thanks for flagging it!

**Review by gjcolombo — APPROVED** [review]

File: lib/propolis/src/block/mod.rs
Comment (pfmooney) [inline lib/propolis/src/block/mod.rs:232]: I'm not wild about there being a default implementation here.  It would be misleading for a block device to accept setting a completion callback, only to never call it for any of its completed requests.

File: lib/propolis/src/block/tracking.rs
Comment (pfmooney) [inline lib/propolis/src/block/tracking.rs:80]: Not required if `with_completion_callback` is eliminated.

File: lib/propolis/src/block/mod.rs
Comment (bnaecker) [reply lib/propolis/src/block/mod.rs:232]: That's fair. How about returning something like an `anyhow::Result<bool>`? That would return whether the callback existed previously in the `Ok(_)` variant, or fail if the device chose not to support it.

Comment (pfmooney) [reply lib/propolis/src/block/mod.rs:232]: > if the device chose not to support it.

All of the existing devices do support it, so I wouldn't worry about that sort of thing for now.  If things change in the future, we can deal with it then.

Comment (bnaecker) [reply lib/propolis/src/block/mod.rs:232]: So to be concrete, you're only arguing that we remove the default impl? Otherwise everything else remains unchanged about this interface?

Comment (pfmooney) [reply lib/propolis/src/block/mod.rs:232]: > you're only arguing that we remove the default impl

yeah, the rest is fine as-is

Comment (bnaecker) [reply lib/propolis/src/block/mod.rs]: Removed the default impl in 8565b0c

File: lib/propolis/src/block/tracking.rs
Comment (bnaecker) [reply lib/propolis/src/block/tracking.rs]: Removed in 8565b0c

File: lib/propolis/src/block/tracking.rs
Comment (bnaecker) [reply lib/propolis/src/block/tracking.rs]: Removed in 8565b0c

**Review by pfmooney — APPROVED** [review]

**Comment by bnaecker** [comment]
Just a note on the sampling interval. I talked in chat with @leftwo, and we decided that the current sampling rate of 10s is sufficient. In this PR, the sampling rate of these statistics is defined by the rate at which oximeter polls -- Propolis only generates sample on a request from oximeter. If we wanted to change this, we have two choices. We could lower that polling rate from oximeter, or we could decouple the rate at which we internally generate samples from the block-dev producer and the rate at which oximeter asks Propolis for samples. That's an architectural change, since we'd need a task that continually generates samples from the block-dev. It's also fairly easy to do, and not necessary right now. We can decide that later without any breaking changes downstream.

**Review by leftwo — APPROVED** [review]

File: Cargo.toml
Comment (bnaecker) [reply Cargo.toml]: Done in 3382496

File: Cargo.toml
Comment (bnaecker) [reply Cargo.toml]: Done in 3382496

**Pull request merged**
